vllm/utils.py

```diff
@@ -510,6 +512,187 @@ def random_uuid() -> str:
+        self._batcher_tasks: dict[tuple, asyncio.Task] = {}
+
+        # Single worker that owns the blocking tokenizer.
+        self._executor = ThreadPoolExecutor(max_workers=1)
+
+    # ------------------------------------------------------------------
+    # Public async API
+    # ------------------------------------------------------------------
+    async def __call__(self, prompt, **kwargs):
+        fut: asyncio.Future = self._loop.create_future()
+        key = self._queue_key("encode", kwargs)
+        queue = self._ensure_queue_and_batcher(self._loop, key)
+        await queue.put(("encode", prompt, kwargs, fut))
+        return await fut
+
+    async def decode(self, token_ids, **kwargs):
+        fut: asyncio.Future = self._loop.create_future()
+        key = self._queue_key("decode", kwargs)
+        queue = self._ensure_queue_and_batcher(self._loop, key)
+        await queue.put(("decode", token_ids, kwargs, fut))
+        return await fut
+
+    # ------------------------------------------------------------------
+    # Internal helpers
+    # ------------------------------------------------------------------
+    def _ensure_queue_and_batcher(self, loop: asyncio.AbstractEventLoop,
+                                  key: tuple):
+        """Return the queue for key, creating queue 
+        and batcher task if needed."""
+        if key not in self._queues:
+            self._queues[key] = asyncio.Queue()
+
+        if key not in self._batcher_tasks or self._batcher_tasks[key].done():
+            self._batcher_tasks[key] = loop.create_task(
+                self._batch_loop(self._queues[key]))
+
+        return self._queues[key]
+
+    async def _batch_loop(self, queue: asyncio.Queue):
+        """Runs until the queue stays empty long enough to finish."""
+        while True:
+            first_item = await queue.get()
+            batch = [first_item]
+            deadline = self._loop.time() + self.batch_wait_timeout_s
+
+            while len(batch) < self.max_batch_size:
+                timeout = deadline - self._loop.time()
+                if timeout <= 0:
+                    break
+                try:
+                    item = await asyncio.wait_for(queue.get(), timeout)
+                    batch.append(item)
+                except asyncio.TimeoutError:
+                    break
+
+            # Split by op type
+            encode_items = [it for it in batch if it[0] == "encode"]
+            decode_items = [it for it in batch if it[0] == "decode"]
+
+            # ---------------- Encode ----------------
+            if encode_items:
+                prompts = [it[1] for it in encode_items]
+                kwargs_list = [it[2] for it in encode_items]
+
+                # If every request uses identical kwargs we can run a single
+                # batched tokenizer call for a big speed-up.
+                can_batch = all(kw == kwargs_list[0] for kw in kwargs_list)
```

Comment:
Similarly I think this can be an arg that we pass in.. since in all but one of the cases the kwargs will be the same (by design). Part of the reason for doing that is to avoid this overhead.

And no need to build the full kwargs list in this case.